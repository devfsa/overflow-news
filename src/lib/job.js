function crawlFeed(URL, callback) {
    const FeedParser = require('feedparser');
    const request = require('request');

    const feedparser = new FeedParser();
    let posts = [];

    request(URL)
        .on('error', function (error) {
            console.log('\x1b[31m[ERROR]\x1b[0m', error.message);
            callback(error);
        })
        .on('response', function (response) {
            this.pipe(feedparser);

            feedparser.on('error', function (error) {
                console.log('\x1b[31m[ERROR]\x1b[0m', error.message);
            });

            feedparser.on('readable', function () {
                let post = null;

                while (post = this.read()) {
                    posts.push(post);
                }
            });

            feedparser.on('end', function () {
                callback(null, posts);
            });
        });
}

function fetchLatestPosts() {
    const Feed = require('../models/feed');

    Feed.find({}, function (error, feeds) {
        if (error) {
            console.log('\x1b[31m[ERROR]\x1b[0m', error.message);
        } else {
            const Queue = require('bee-queue');
            const queue = new Queue(process.env.CRAWL_QUEUE);

            feeds.forEach(function (feed) {
                queue.createJob(feed).save();
            });

            console.log('\x1b[34m[INFO]\x1b[0m', `${feeds.length} feeds to crawl`);
        }
    });
}

module.exports = {
    crawlFeed,
    fetchLatestPosts
};