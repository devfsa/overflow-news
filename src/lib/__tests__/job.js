/* eslint-env jest, browser */

// mocking
jest.mock('request');
jest.mock('feedparser');
import requestMock from 'request';
import fpMock from 'feedparser';

import { crawlFeed } from '../job'

describe('crawlFeed', () => {
  beforeEach(() => {
    fpMock.mockClear();
  });

  it('request calls', () => {
    const req = { on: jest.fn() };
    req.on.mockReturnValue(req);
    requestMock.mockReturnValue(req);

    const url = `some.url/${Math.random()}`;
    crawlFeed(url);

    expect(requestMock).toHaveBeenCalledTimes(1);
    expect(requestMock).toHaveBeenCalledWith(url);

    expect(req.on.mock.calls).toMatchSnapshot();
  });

  it('request error handler', () => {
    const req = { on: jest.fn() };
    req.on.mockReturnThis();
    requestMock.mockReturnValue(req);

    const callback = jest.fn();
    const url = `some.url/${Math.random()}`;
    crawlFeed(url, callback);

    const forcedError = new Error('forcing');
    const handler = req.on.mock.calls.filter(call => call[0] === 'error')[0][1];
    handler(forcedError);

    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(forcedError);
  });

  it('request response handler', () => {
    const fp = { on: jest.fn() };
    const req = { on: jest.fn(), pipe: jest.fn() };
    fp.on.mockReturnThis();
    req.on.mockReturnThis();
    requestMock.mockReturnValue(req);
    fpMock.mockImplementation(() => fp);

    const url = `some.url/${Math.random()}`;
    crawlFeed(url);

    const handler = req.on.mock.calls.filter(call => call[0] === 'response')[0][1];
    handler.bind(req)();

    expect(req.pipe).toHaveBeenCalledTimes(1);
    expect(req.pipe).toHaveBeenCalledWith(fp);

    expect(fp.on.mock.calls).toMatchSnapshot();
  });

  it('feedparser error handler', () => {
    const fp = { on: jest.fn() };
    const req = { on: jest.fn(), pipe: jest.fn() };
    fp.on.mockReturnThis();
    req.on.mockReturnThis();
    requestMock.mockReturnValue(req);
    fpMock.mockImplementation(() => fp);

    const callback = jest.fn();
    const url = `some.url/${Math.random()}`;
    crawlFeed(url, callback);

    const reqHandler = req.on.mock.calls.filter(call => call[0] === 'response')[0][1];
    reqHandler.bind(req)();

    const forcedError = new Error('forcing');
    const fpHandler = fp.on.mock.calls.filter(call => call[0] === 'error')[0][1];
    fpHandler(forcedError);

    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(forcedError);
  });

  it('feedparser success handler', () => {
    const fp = { on: jest.fn() };
    const req = { on: jest.fn(), pipe: jest.fn() };
    fp.on.mockReturnThis();
    req.on.mockReturnThis();
    requestMock.mockReturnValue(req);
    fpMock.mockImplementation(() => fp);

    const callback = jest.fn();
    const url = `some.url/${Math.random()}`;
    crawlFeed(url, callback);

    const reqHandler = req.on.mock.calls.filter(call => call[0] === 'response')[0][1];
    reqHandler.bind(req)();

    const [fpReadable, fpEnd] = fp.on.mock.calls
        .filter(call => call[0] !== 'error')
        .map(params => params[1]);

    const post = Math.random();
    const fakeReadable = { read: jest.fn() };
    fakeReadable.read.mockReturnValueOnce(post);
    fakeReadable.read.mockReturnValueOnce(null);
    fpReadable.bind(fakeReadable)();
    fpEnd();

    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(null, [post]);
  });
});
