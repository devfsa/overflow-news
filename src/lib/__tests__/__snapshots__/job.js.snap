// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crawlFeed assert request calls 1`] = `
Array [
  Array [
    "error",
    [Function],
  ],
  Array [
    "response",
    [Function],
  ],
]
`;

exports[`crawlFeed assert response handler 1`] = `
Array [
  Array [
    "error",
    [Function],
  ],
  Array [
    "readable",
    [Function],
  ],
  Array [
    "end",
    [Function],
  ],
]
`;
